File: src/commands/sync.ts
import { Args, Command } from "@effect/cli"
import { Console, Effect } from "effect"
import { ConfigService } from "../services/config.js"
import { SyncService } from "../services/sync.js"
import { bold, dim, green, yellow, red } from "../fmt.js"

// ---------------------------------------------------------------------------
// ship sync <project>
// ---------------------------------------------------------------------------

const projectArg = Args.text({ name: "project" })

export const syncCommand = Command.make(
  "sync",
  { project: projectArg },
  ({ project }) =>
    Effect.gen(function* () {
      const config = yield* ConfigService
      const sync = yield* SyncService

      const projectConfig = yield* config.getProject(project)

      yield* Console.log("")
      yield* Console.log(`  Syncing ${bold(project)}...`)

      const result = yield* sync.sync(projectConfig)

      yield* Console.log(`  ${green("✓")} Fetched        origin`)

      if (result.pulled && result.headMoved) {
        yield* Console.log(`  ${green("✓")} Pulled         main ${dim("(fast-forward)")}`)
      } else if (result.pulled) {
        yield* Console.log(`  ${dim("  ·")} Pulled         ${dim("already up to date")}`)
      } else if (result.skippedPull) {
        yield* Console.log(`  ${yellow("⚠")} Skipped pull   ${dim(result.skippedPull)}`)
      }

      if (result.installed) {
        yield* Console.log(`  ${green("✓")} Dependencies   installed`)
      }
      if (result.migrated) {
        yield* Console.log(`  ${green("✓")} Migrations     applied ${dim(`(${projectConfig.database.source})`)}`)
      }

      yield* Console.log("")
    }).pipe(
      Effect.catchAll((e) =>
        Console.error(`\n  ${red("Error:")} ${e.message}\n`)
      )
    )
)

File: src/commands/sync.rs
use crate::errors::Result;
use crate::fmt::{bold, dim, green, red, yellow};
use crate::services::{config, sync};

// ---------------------------------------------------------------------------
// ship sync <project>
// ---------------------------------------------------------------------------

pub fn run(project: String) {
    if let Err(e) = run_inner(&project) {
        eprintln!("\n  {} {}\n", red("Error:"), e);
    }
}

fn run_inner(project: &str) -> Result<()> {
    let project_config = config::get_project(project)?;

    println!();
    println!("  Syncing {}...", bold(project));

    let result = sync::sync(&project_config, None)?;

    println!("  {} Fetched        origin", green("✓"));

    if result.pulled && result.head_moved {
        println!("  {} Pulled         main {}", green("✓"), dim("(fast-forward)"));
    } else if result.pulled {
        println!("  {} Pulled         {}", dim("  ·"), dim("already up to date"));
    } else if let Some(skipped) = &result.skipped_pull {
        println!("  {} Skipped pull   {}", yellow("⚠"), dim(skipped));
    }

    if result.installed {
        println!("  {} Dependencies   installed", green("✓"));
    }
    if result.migrated {
        println!(
            "  {} Migrations     applied {}",
            green("✓"),
            dim(format!("({})", project_config.database.source))
        );
    }

    println!();
    Ok(())
}
